python/lib/core/dmod/core/common/collections/catalog.py

```diff
@@ -0,0 +1,651 @@
+
+        Returns:
+            The data from the removed entry
+        """
+        if resolve is None:
+            resolve = True
+
+        if resolve:
+            await self.commit()
+
+        try:
+            del self[key]
+        finally:
+            await self.commit()
+
+        return None
+
+    def __delitem__(self, key: str) -> None:
+        """
+        Remove an entry from the cache
+
+        Triggered asynchronous tasks will be stored for later completion
+
+        Args:
+            key: The identifier for the item to remove
+        """
+        try:
+            self.lock()
+            super().__delitem__(key)
+        finally:
+            self.__update_earliest()
+            self.release()
+
+    def __getitem__(self, key: str) -> typing.Union[HashableTypeT, None]:
+        """
+        Get data from the Cache
+
+        Triggered asynchronous tasks will be stored for later completion
+
+        Args:
+            key: The identifier for the entry to get data from
+
+        Returns:
+            The data from within the entry if it was present
+        """
+        try:
+            self.lock()
+
+            if key not in self:
+                raise KeyError(f"Key: '{key}' not found in instance of {self.__class__.__name__}")
+
+            entry = super().get(key)
+
+            if isinstance(entry, CatalogEntry):
+                entry.touch()
+                entry = entry.data
+
+            return entry
+        finally:
+            self.__update_earliest()
+            self.release()
+
+    def get(self, key: str, default: T = None) -> typing.Union[HashableTypeT, T]:
+        """
+        Get data from the cache or a default if it is not present
+
+        Triggered asynchronous tasks will be stored for later completion
+
+        Args:
+            key: The identifier for the entry to get data from
+            default: A value to return if the entry was not present
+
+        Returns:
+            The value of the entry or the default
+        """
+        return self[key] if key in self else default
+
+    async def get_async(
+        self,
+        key: str,
+        default: T = None,
+        resolve: bool = None
+    ) -> typing.Union[bytes, T]:
+        """
+        Get data from the cache or a default if it is not present
+
+        All triggered asynchronous tasks will be completed prior to return
+
+        Args:
+            key: The identifier for the entry to get data from
+            default: A value to return if the entry was not present
+            resolve: Whether to complete all leftover tasks prior to retrieving data
+
+        Returns:
+            The value of the entry or the default
+        """
+        if resolve is None:
+            resolve = True
+
+        if resolve:
+            await self.commit()
+
+        try:
+            return self.get(key, default)
+        finally:
+            await self.commit()
+
+    def lock(self):
```

Comment:
If this is going to act as a base class, can you inherit from `abc` and mark that this is abstract please?